File: RESTFul/Payment/Java/app1/src/main/java/com/att/api/payment/controller/PaymentController.java
package com.att.api.payment.controller;

import java.io.IOException;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.json.JSONObject;

import com.att.api.controller.APIController;
import com.att.api.oauth.OAuthToken;
import com.att.api.payment.file.PaymentFileHandler;
import com.att.api.payment.file.SubscriptionEntry;
import com.att.api.payment.file.TransactionEntry;
import com.att.api.payment.model.AppCategory;
import com.att.api.payment.model.ConfigBean;
import com.att.api.payment.model.Notary;
import com.att.api.payment.model.NotificationPool;
import com.att.api.payment.model.Subscription;
import com.att.api.payment.model.Transaction;
import com.att.api.payment.service.NotaryService;
import com.att.api.payment.service.PaymentService;
import com.att.api.rest.RESTException;

public class PaymentController extends APIController {

    private static final long serialVersionUID = 5677394140665947979L;
    private static PaymentFileHandler transactionFile = null;
    private static PaymentFileHandler subscriptionFile = null;

    private Map<String, String> JSONToMap(JSONObject obj) {
        Map<String, String> map = new TreeMap<String, String>();
        @SuppressWarnings("rawtypes")
        Iterator keys = obj.keys();
        while (keys.hasNext()) {
            String key = (String) keys.next();
            map.put(key, obj.getString(key));
        }
        return map;
    }

    private List<TransactionEntry> createTransactionEntries(List<JSONObject> list){
        List<TransactionEntry> entries = new ArrayList<TransactionEntry>();
        for(JSONObject json : list){
            entries.add(new TransactionEntry(json));
        }
        return entries;
    }

    private List<SubscriptionEntry> createSubscriptionEntries(List<JSONObject> list){
        List<SubscriptionEntry> entries = new ArrayList<SubscriptionEntry>();
        for(JSONObject json : list){
            entries.add(new SubscriptionEntry(json));
        }
        return entries;
    }

    private void createFileHandler() {
        try {
            if (transactionFile == null || subscriptionFile == null) {
                transactionFile = new PaymentFileHandler("trans", "db");
                subscriptionFile = new PaymentFileHandler("subs", "db");
            }
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    private Notary getTransactionNotary(double price) throws RESTException {
        String transId = "J" + System.currentTimeMillis();
        String prodId = appConfig.getProperty("merchantProductId");
        String redirectUrl = appConfig.getProperty("paymentRedirectURL");
        Transaction trans = new Transaction.Builder(price, AppCategory.APP_OTHER,
                "Sample App", transId, prodId, redirectUrl).build();

        NotaryService srvc = new NotaryService(appConfig.getFQDN(), appConfig.getClientId(), 
                appConfig.getClientSecret());
        return srvc.getTransactionNotary(trans);
    }

    private Notary getSubscriptionNotary(double price) throws RESTException {
        String transId = "J" + System.currentTimeMillis();
        String mpid = appConfig.getProperty("merchantProductId");
        String redirectUrl = appConfig.getProperty("paymentRedirectURL");
        String merchSubIdList = appConfig.getProperty("merchantSubscriptionId");

        Subscription sub = new Subscription.Builder(price, AppCategory.APP_OTHER,
                "Sample App", transId, mpid, redirectUrl, merchSubIdList).build();

        NotaryService srvc = new NotaryService(appConfig.getFQDN(), appConfig.getClientId(), 
                appConfig.getClientSecret());
        return srvc.getSubscriptionNotary(sub);
    }

    private JSONObject getTransactionInfo(String type, String value)
        throws RESTException {

        //TODO: update this to enums
        String transType = "TransactionAuthCode";
        if (type.equals("2")) {
            transType = "TransactionId";
        } else if (type.equals("3")) {
            transType = "MerchantTransactionId";
        }

        PaymentService srvc = new PaymentService(appConfig.getFQDN(),
                getFileToken());
        return srvc.getTransactionStatus(transType, value);
    }

    private JSONObject getSubscriptionInfo(String type, String value)
        throws RESTException {

        String subType = "SubscriptionAuthCode";
        if (type.equals("2")) {
            subType = "SubscriptionId";
        } else if (type.equals("3")) {
            subType = "MerchantTransactionId";
        }

        PaymentService srvc = new PaymentService(appConfig.getFQDN(),
                getFileToken());
        return srvc.getSubscriptionStatus(subType, value);
    }

    private void handleTransactionAuthCode(HttpServletRequest request) {
        try {
            if (request.getParameter("TransactionAuthCode") == null) {
                return;
            }

            String authCode = (String) request.getParameter("TransactionAuthCode");
            JSONObject info = this.getTransactionInfo("1", authCode);
            transactionFile.addTransactionEntry(info, authCode);
            request.setAttribute("showTrans", true);
            Map<String, String> m = this.JSONToMap(info);
            m.remove("authCode");
            request.setAttribute("resultTransInfo", m);
        } catch (Exception e) {
            request.setAttribute("showTrans", true);
            request.setAttribute("errorTransAuthCode", e.getMessage());
        }
    }

    private void handleSubscriptionAuthCode(HttpServletRequest request){
        try {
            if (request.getParameter("SubscriptionAuthCode") == null) {
                return;
            }
            String authCode = (String) request.getParameter("SubscriptionAuthCode");
            JSONObject info = this.getSubscriptionInfo("1", authCode);
            subscriptionFile.addTransactionEntry(info, authCode);
            request.setAttribute("showSub", true);
            Map<String, String> m = this.JSONToMap(info);
            m.remove("authCode");
            request.setAttribute("resultSubInfo", m);
        } catch (Exception e) {
            request.setAttribute("showSub", true);
            request.setAttribute("errorSubAuthCode", e.getMessage());
        }

    }

    private void handleGetTransactionStatus(HttpServletRequest request) {
        try {
            String[] names = { "getTransactionMTID", "getTransactionAuthCode",
                "getTransactionTID" };
            String[] types = { "3", "1", "2" };

            for (int i = 0; i < names.length; ++i) {
                String name = names[i];
                if (request.getParameter(name) != null) {
                    String value = request.getParameter(name);
                    JSONObject info = this.getTransactionInfo(types[i], value);
                    request.setAttribute("showTrans", true);
                    request.setAttribute("resultTransInfo", this.JSONToMap(info));
                    return;
                }
            }
        } catch (Exception e) {
            request.setAttribute("showTrans", true);
            request.setAttribute("errorTransInfo", e.getMessage());
        }
    }

    private void handleGetSubscriptionStatus(HttpServletRequest request) {
        try {
            String[] names = { "getSubscriptionMTID", "getSubscriptionAuthCode",
                "getSubscriptionTID" };
            String[] types = { "3", "1", "2" };

            for (int i = 0; i < names.length; ++i) {
                String name = names[i];
                if (request.getParameter(name) != null) {
                    String value = request.getParameter(name);
                    JSONObject info = this.getSubscriptionInfo(types[i], value);
                    request.setAttribute("showSub", true);
                    request.setAttribute("resultSubInfo", this.JSONToMap(info));
                    return;
                }
            }
        } catch (Exception e) {
            request.setAttribute("showSub", true);
            request.setAttribute("errorSubInfo", e.getMessage());
        }
    }

    private void handleGetSubscriptionDetail(HttpServletRequest request) {
        try {
            String mid = request.getParameter("getSDetailsMSID");

            if (mid == null)
                return;

            List<SubscriptionEntry> subList = createSubscriptionEntries(subscriptionFile.getTransactionEntrys());

            String merch = mid;
            String consumer = "";

            for (SubscriptionEntry sub : subList){
                if (sub.getMerchantSubId().equals(mid))
                    consumer = sub.getConsumerId();
            }

            OAuthToken token = this.getFileToken();

            JSONObject info = new PaymentService(appConfig.getFQDN(), 
                    token).getSubscriptionDetails(merch, consumer);

            request.setAttribute("showSub", true);
            request.setAttribute("resultSubDetail", this.JSONToMap(info));
        } catch (Exception e) {
            request.setAttribute("showSub", true);
            request.setAttribute("errorSubDetail", e.getMessage());
        }
    }

    private boolean handleNewTransaction(HttpServletRequest request,
            HttpServletResponse response) {

        try {
            if (request.getParameter("newTransaction") == null) {
                // no redirect
                return false;
            }

            String prod = (String) request.getParameter("product");
            String cost = appConfig.getProperty("minTransactionAmount");
            if (prod.equals("2")) {
                cost = appConfig.getProperty("maxTransactionAmount");
            }

            Notary notary = this.getTransactionNotary(Double.valueOf(cost));
            request.getSession().setAttribute("notary", notary);

            String URL = PaymentService.getNewTransactionURL(
                    appConfig.getFQDN(), appConfig.getClientId(), notary);
            response.sendRedirect(URL);
            // redirect required
            return true;
        } catch (Exception e) {
            request.setAttribute("errorNewTrans", e.getMessage());
            return false;
        }
    }

    private boolean handleNewSubscription(HttpServletRequest request,
            HttpServletResponse response) {

        try {
            if (request.getParameter("newSubscription") == null) {
                // no redirect
                return false;
            }

            String prod = (String) request.getParameter("product");
            String cost = appConfig.getProperty("minSubscriptionAmount");
            if (prod.equals("2")) {
                cost = appConfig.getProperty("maxSubscriptionAmount");
            }

            Notary notary = this.getSubscriptionNotary(Double.valueOf(cost));
            request.getSession().setAttribute("notary", notary);

            String URL = PaymentService.getNewSubscriptionURL(
                    appConfig.getFQDN(), appConfig.getClientId(), notary);
            response.sendRedirect(URL);
            // redirect required
            return true;
        } catch (Exception e) {
            request.setAttribute("errorNewSub", e.getMessage());
            return false;
        }
    }

    private void handleTransRefund(HttpServletRequest request) {
        try {
            if (request.getParameter("refundTransactionId") == null) {
                return;
            }
            String tid = (String) request.getParameter("refundTransactionId");

            PaymentService srvc = new PaymentService(appConfig.getFQDN(), 
                    this.getFileToken());

            JSONObject info = srvc.refundTransaction(tid, 
                    "Sample App Test", 1);

            request.setAttribute("showTrans", true);
            request.setAttribute("resultTRefund", this.JSONToMap(info));
        } catch (Exception e) {
            request.setAttribute("showTrans", true);
            request.setAttribute("errorTRefund", e.getMessage());
        }
    }

    private void handleSubscriptionRefund(HttpServletRequest request){
        try {
            if (request.getParameter("refundSubscriptionId") == null) {
                return;
            }
            String sid = (String) request.getParameter("refundSubscriptionId");

            PaymentService srvc = new PaymentService(appConfig.getFQDN(), 
                    getFileToken());

            JSONObject info = srvc.refundSubscription(sid, "Sample App Test", 1);

            request.setAttribute("showSub", true);
            request.setAttribute("resultSRefund", this.JSONToMap(info));
        } catch (Exception e) {
            request.setAttribute("showSub", true);
            request.setAttribute("errorSRefund", e.getMessage());
        }
    }

    private void handleSubscriptionCancel(HttpServletRequest request){
        try {
            if (request.getParameter("cancelSubscriptionId") == null) {
                return;
            }
            String sid = (String) request.getParameter("cancelSubscriptionId");

            PaymentService srvc = new PaymentService(appConfig.getFQDN(), 
                    getFileToken());

            JSONObject info = srvc.cancelSubscription(sid, "Sample App Test", 1);
            request.setAttribute("showSub", true);
            request.setAttribute("resultSCancel", this.JSONToMap(info));
        } catch (Exception e) {
            request.setAttribute("showSub", true);
            request.setAttribute("errorSCancel", e.getMessage());
        }
    }

    private void handleNotary(HttpServletRequest request) {
        try {
            if (request.getParameter("signPayload") == null) {
                Notary notary = (Notary) request.getSession().getAttribute("notary");
                if (notary != null){
                    request.setAttribute("showNotary", true);
                    request.setAttribute("notaryPayload", notary.getPayload().trim());
                    request.setAttribute("resultNotaryDoc", notary.getSignedDocument());
                    request.setAttribute("resultNotarySig", notary.getSignature());
                }
                return;
            }
            String payload = (String) request.getParameter("payload");

            NotaryService srvc = new NotaryService(appConfig.getFQDN(), 
                    appConfig.getClientId(), appConfig.getClientSecret());

            Notary notary = srvc.getNotary(payload);
            request.setAttribute("showNotary", true);
            request.setAttribute("notaryPayload", notary.getPayload().trim());
            request.setAttribute("resultNotaryDoc", notary.getSignedDocument());
            request.setAttribute("resultNotarySig", notary.getSignature());
        } catch (Exception e) {
            request.setAttribute("showNotary", true);
            request.setAttribute("errorNotary", e.getMessage());
        }
    }

    private void handleNotifications(HttpServletRequest request) {
        NotificationPool pool = NotificationPool.getInstance();
        request.setAttribute("notifications", pool.getNotifications());
        if (request.getParameter("refreshNotifications") != null)
            request.setAttribute("showNote", true);
    }

    private void setFileResults(HttpServletRequest request) {
        this.createFileHandler();
        try {
            request.setAttribute("transactions", createTransactionEntries(
                        transactionFile.getTransactionEntrys()));

            request.setAttribute("subscriptions", createSubscriptionEntries(
                        subscriptionFile.getTransactionEntrys()));
        } catch (IOException e) {
            // don't handle, print stack trace
            e.printStackTrace();
        } catch (ParseException e){
            // don't handle, print stack trace
            e.printStackTrace();
        }
    }

    public void doPost(HttpServletRequest request, HttpServletResponse 
            response) throws ServletException, IOException {
        doGet(request,response);
            }

    public void doGet(HttpServletRequest request, HttpServletResponse 
            response) throws ServletException, IOException {

        request.setAttribute("cfg", new ConfigBean());

        this.handleNotifications(request);

        // handle transactions
        if (this.handleNewTransaction(request, response)) {
            return;
        }
        this.handleTransactionAuthCode(request); 
        this.handleGetTransactionStatus(request);
        this.handleTransRefund(request);

        // handle notary
        this.handleNotary(request);

        //handle subscriptions
        if (this.handleNewSubscription(request, response)) {
            return;
        }
        this.handleSubscriptionAuthCode(request);
        this.handleGetSubscriptionStatus(request);
        this.handleGetSubscriptionDetail(request);
        this.handleSubscriptionRefund(request);
        this.handleSubscriptionCancel(request);

        this.setFileResults(request);

        // forward to view
        String forwardStr = "/WEB-INF/payment.jsp";
        request.getRequestDispatcher(forwardStr).forward(request, response);
            }
}

File: RESTFul/Payment/PHP/app1/src/Controller/PaymentController.php
<?php

require_once __DIR__ . '/../../lib/Controller/APIController.php';
require_once __DIR__ . '/../../lib/Payment/PaymentFileHandler.php';
require_once __DIR__ . '/../../lib/Payment/PaymentService.php';
require_once __DIR__ . '/../../lib/Util/Util.php';

class PaymentController extends APIController {
    // TODO: Move string indices to constants

    private $_minTransValue;
    private $_maxTransValue;
    private $_minSubValue;
    private $_maxSubValue;
    private $_redirectURL;

    private $_arrIds;

    private $_fileHandler;

    private function getSubscriptionDetails($merchantSID, $consumerId) {
        $paymentReq = $this->createPaymentService();
        return $paymentReq->getSubscriptionDetails($merchantSID, $consumerId);
    }

    private function createPaymentService() {
        $srvc = new PaymentService($this->apiFQDN, $this->getFileToken());
        return $srvc;
    }

    private function getSubscriptionInfo($type, $value) {
        // TODO: Change subURL varname to a better name 
        $subURL = 'SubscriptionAuthCode';
        if (strcmp($type, '2') == 0) {
            $subURL = 'MerchantTransactionId';
        } else if (strcmp($type, '3') == 0) {
            $subURL = 'SubscriptionId';
        }

        $paymentReq = $this->createPaymentService();
        $info = $paymentReq->getSubscriptionStatus($subURL, $value);
        return $info;
    }

    private function getTransactionInfo($type, $value) {
        // TODO: Change subURL varname to a better name 
        $subURL = 'TransactionAuthCode';
        if (strcmp($type, '2') == 0) {
            $subURL = 'TransactionId';
        } else if (strcmp($type, '3') == 0) {
            $subURL = 'MerchantTransactionId';
        }

        $paymentReq = $this->createPaymentService(); 
        $info = $paymentReq->getTransactionStatus($subURL, $value);
        return $info;
    }

    private function getSubscriptionNotary($price) {
        $subArgs = new SubscriptionNotaryArguments();
        $subArgs
            ->setAmount($price)
            ->setCategory(NotaryArguments::CATEGORY_APPLICATION_OTHER)
            ->setDescription('SAMPLE APP')
            ->setMerchantTransactionId('mtid' . time())
            ->setMerchantProductId('mpid' . time())
            ->setMerchantRedirectUrl($this->_redirectURL);

        $request = new NotaryService($this->apiFQDN, $this->clientId, 
                $this->clientSecret);
        return $request->getSubscriptionNotary($subArgs);
    }

    private function getTransactionNotary($price) {
        $transArgs = new TransactionNotaryArguments();
        $transArgs
            ->setAmount($price)
            ->setCategory(NotaryArguments::CATEGORY_APPLICATION_OTHER)
            ->setDescription('SAMPLE APP')
            ->setMerchantTransactionId('mtid' . time())
            ->setMerchantProductId('mpid' . time())
            ->setMerchantRedirectUrl($this->_redirectURL);

        $request = new NotaryService($this->apiFQDN, $this->clientId, 
                $this->clientSecret);
        return $request->getTransactionNotary($transArgs);
    }

    private function handleNotary() {
        try {
            if (!isset($_REQUEST['signPayload'])) {
                return;
            }

            $payload = $_REQUEST['payload'];
            $req = new NotaryService($this->apiFQDN, $this->clientId, 
                    $this->clientSecret);
            $this->results['notary'] = $req->getNotary($payload);
        } catch (Exception $e) {
            $this->errors['notary'] = $e->getMessage();
        }
    }

    private function handleNewSubscription() {
        try {
            if (!isset($_REQUEST['newSubscription'])) {
                return;
            }

            $cost = $this->_minSubValue;
            if(strcmp($_REQUEST['product'], '2') == 0) {
                $cost = $this->_maxSubValue;
            }
            $notary = $this->getSubscriptionNotary($cost);
            $_SESSION['notary'] = serialize($notary);

            // redirect
            PaymentService::newSubscription($this->apiFQDN, $this->clientId, 
                    $notary); 
        } catch (Exception $e) {
            $this->errors['newSub'] = $e->getMessage();
        }
    }
    private function handleNewTransaction() {
        try {
            if (!isset($_REQUEST['newTransaction'])) {
                return;
            }

            $cost = $this->_minTransValue;
            if(strcmp($_REQUEST['product'], '2') == 0) {
                $cost = $this->_maxTransValue;
            }
            $notary = $this->getTransactionNotary($cost);
            $_SESSION['notary'] = serialize($notary);

            // redirect
            PaymentService::newTransaction($this->apiFQDN, $this->clientId, 
                    $notary); 
        } catch (Exception $e) {
            $this->errors['newTrans'] = $e->getMessage();
        }
    }

    private function handleSubscriptionAuthCode() {
        try {
            if (!isset($_REQUEST['SubscriptionAuthCode'])) {
                return;
            }

            $authCode = $_REQUEST['SubscriptionAuthCode'];
            $info = $this->getSubscriptionInfo('1', $authCode);
            $merchantTransId = $info['MerchantTransactionId'];
            $subId = $info['SubscriptionId'];
            $consumerId = $info['ConsumerId'];
            $merchantSubId = $info['MerchantSubscriptionId'];
            $this->results['subInfo'] = $info;
            $this->_fileHandler->addSubscriptionInfo($subId, $merchantTransId, 
                    $authCode, $consumerId, $merchantSubId);
        } catch (Exception $e) {
            $this->errors['subInfo'] = $e->getMessage();
        }
    }

    private function handleTransactionAuthCode() {
        try {
            if (!isset($_REQUEST['TransactionAuthCode'])) {
                return;
            }
            
            $authCode = $_REQUEST['TransactionAuthCode'];
            $info = $this->getTransactionInfo('1', $authCode);

            $merchantTransId = $info['MerchantTransactionId'];
            $transId = $info['TransactionId'];
            $this->results['transInfo'] = $info;
            $this->_fileHandler->addTransactionInfo(
                    $transId, $merchantTransId, $authCode);
        } catch (Exception $e) {
            $this->errors['transAuthCode'] = $e->getMessage();
        }
    }

    private function handleGetSubscriptionDetails() {
        try {
            if (isset($_REQUEST['getSDetailsMSID'])) {
                $fValues = $this->_fileHandler->getFileValues();
                $merchantSubIds 
                    = $fValues[PaymentFileHandler::INDEX_S_MERCHANT_SUB_ID];

                $merchantSubId = $_REQUEST['getSDetailsMSID'];
                $index = array_search($merchantSubId, $merchantSubIds);

                $consumerIds 
                    = $fValues[PaymentFileHandler::INDEX_S_CONSUMER_ID];
                $consumerId = $consumerIds[$index];

                $dinfo = $this->getSubscriptionDetails($merchantSubId, 
                        $consumerId);
                $this->results['subDetails'] = $dinfo;
            } else if (isset($_REQUEST['getSDetailsConsumerId'])) {
                $fValues = $this->_fileHandler->getFileValues();
                $merchantSubIds 
                    = $fValues[PaymentFileHandler::INDEX_S_MERCHANT_SUB_ID];
                $consumerIds 
                    = $fValues[PaymentFileHandler::INDEX_S_CONSUMER_ID];

                $consumerId = $_REQUEST['getSDetailsConsumerId'];
                $index = array_search($consumerId, $consumerIds);

                $merchantSubId = $merchantSubIds[$index];

                $dinfo = $this->getSubscriptionDetails($merchantSubId,
                        $consumerId);
                $this->results['subDetails'] = $dinfo;
            }
        } catch (Exception $e) {
            $this->errors['subDetails'] = $e->getMessage();
        }
    }

    private function handleGetSubscriptionStatus() {
        try {
            $names = array('getSubscriptionMTID', 'getSubscriptionAuthCode', 
                    'getSubscriptionTID');
            $types = array('2', '1', '3');
            for ($i = 0; $i < count($names); ++$i) {
                $name = $names[$i];
                if (isset($_REQUEST[$name])) {
                    $value = $_REQUEST[$name];
                    $info = $this->getSubscriptionInfo($types[$i], $value);
                    $this->results['subInfo'] = $info;
                    return;
                }
            }
        } catch (Exception $e) {
            $this->errors['subInfo'] = $e->getMessage();
        }
    }

    private function handleGetTransactionStatus() {
        try {
            $names = array('getTransactionMTID', 'getTransactionAuthCode', 
                    'getTransactionTID');
            $types = array('3', '1', '2');
            for ($i = 0; $i < count($names); ++$i) {
                $name = $names[$i];
                if (isset($_REQUEST[$name])) {
                    $value = $_REQUEST[$name];
                    $info = $this->getTransactionInfo($types[$i], $value);
                    $this->results['transInfo'] = $info;
                    return;
                }
            }
        } catch (Exception $e) {
            $this->errors['transInfo'] = $e->getMessage();
        }
    }

    private function handleSubRefund() {
        try {
            if (!isset($_REQUEST['refundSubscriptionId'])) {
                return;
            }
            $sid = $_REQUEST['refundSubscriptionId'];

            $request = $this->createPaymentService(); 
            $rinfo = $request->refundSubscription($sid, 'Sample App Test');
            $this->results['s_refund'] = $rinfo;
        } catch (Exception $e) {
            $this->errors['s_refund'] = $e->getMessage();
        }
    }

    private function handleSubCancel() {
        try {
            if (!isset($_REQUEST['cancelSubscriptionId'])) {
                return;
            }
            $sid = $_REQUEST['cancelSubscriptionId'];

            $request = $this->createPaymentService(); 
            $rinfo = $request->cancelSubscription($sid, 'Sample App Test');

            $this->results['s_cancel'] = $rinfo;
        } catch (Exception $e) {
            $this->errors['s_cancel'] = $e->getMessage();
        }
    }

    private function handleTransRefund() {
        try {
            if (!isset($_REQUEST['refundTransactionId'])) {
                return;
            }
            $tid = $_REQUEST['refundTransactionId'];

            $request = $this->createPaymentService(); 
            $rinfo = $request->refundTransaction($tid, 'Sample App Test');
            $this->results['t_refund'] = $rinfo;
        } catch (Exception $e) {
            $this->errors['t_refund'] = $e->getMessage();
        }
    }

    public function __construct() {
        parent::__construct();

        require __DIR__ . '/../../config.php';
        $this->_minTransValue = $minTransactionValue;
        $this->_maxTransValue = $maxTransactionValue;
        $this->_minSubValue = $minSubscriptionValue;
        $this->_maxSubValue = $maxSubscriptionValue;
        $this->_redirectURL = $redirect_url;

        $this->_fileHandler = new PaymentFileHandler('saves.db');
    }


    private function setFileResults() {
        $iNames = array('t_merchantTransIds', 't_authCodes', 't_transIds',
                's_merchantTransIds', 's_authCodes', 's_subIds', 
                's_consumerIds', 's_merchantSubIds');
        $pVals = array(
                PaymentFileHandler::INDEX_MERCHANT_TRANS_ID, 
                PaymentFileHandler::INDEX_AUTH_CODE, 
                PaymentFileHandler::INDEX_TRANS_ID, 
                PaymentFileHandler::INDEX_S_MERCHANT_TRANS_ID, 
                PaymentFileHandler::INDEX_S_AUTH_CODE, 
                PaymentFileHandler::INDEX_SUB_ID, 
                PaymentFileHandler::INDEX_S_CONSUMER_ID,
                PaymentFileHandler::INDEX_S_MERCHANT_SUB_ID
                );
        $fValues = $this->_fileHandler->getFileValues();
        for ($i = 0; $i < count($iNames); ++$i) {
            $iName = $iNames[$i];
            $pVal = $pVals[$i];
            $this->results[$iName] = $fValues[$pVal];
        }
    }
                
    public function handleRequest() {
        // handle transactions
        $this->handleNewTransaction();
        $this->handleTransactionAuthCode();
        $this->handleGetTransactionStatus();
        $this->handleTransRefund();

        // handle subscriptions
        $this->handleNewSubscription();
        $this->handleSubscriptionAuthCode();
        $this->handleGetSubscriptionStatus();
        $this->handleGetSubscriptionDetails();
        $this->handleSubCancel();
        $this->handleSubRefund();

        // handle part notary
        $this->handleNotary();

        // load file contents
        $this->setFileResults();
    
        // notification information
        $this->results['notifications'] 
            = FileUtil::loadArray('notifications.db');

        // notary
        if (isset($_SESSION['notary'])) {
            $this->results['notary'] = unserialize($_SESSION['notary']);
        }
    }

    public function showTransaction() {
        return isset($this->results['transInfo']) 
            || isset($this->errors['transInfo'])
            || isset($this->results['t_refund'])
            || isset($this->errors['t_refund']);
    }

    public function showSubscription() {
        return isset($this->results['subInfo']) 
            || isset($this->errors['subInfo'])
            || isset($this->results['subDetails'])
            || isset($this->errors['subDetails'])
            || isset($this->results['s_refund'])
            || isset($this->errors['s_refund'])
            || isset($this->results['s_cancel'])
            || isset($this->errors['s_cancel']);
    }

    public function showNotary() {
        return isset($this->results['notary']) 
            || isset($this->errors['notary']);
    }

    public function showNotifications() {
        return isset($_REQUEST['refreshNotifications']);
    }
      
}
?>
